numpy/core/defchararray.py

```diff
@@ -57,7 +57,7 @@ def _is_unicode(arr):
     return False
 
 
-def _to_string_or_unicode_array(result, output_dtype_like=None):
+def _to_bytes_or_str_array(result, output_dtype_like=None):
```

Comment:
Maybe we should have `_to_string_or_unicode_array` be an wrapper of `_to_bytes_or_str_array` that emits a `DeprecationWarning`?

Reply:
It starts with `_` so it is a module private function, not to be called outside the module. I realize this is a convention, not enforced by the interpreter.

Do you consider it to be part of the numpy API?

Reply:
Since this is python, everything is public API. If possible, it would be nice to those few people using this hidden function not to have it suddenly disappear without warning. I think in this case the effort to add a small wrapper would be worth it.

Reply:
Yes, I understand that, module private functions starting with `_` are a convention. I may be mistaken, but I haven't seen a DeprecationWarning for a module private function elsewhere in Numpy. Perhaps because such a function has not been deprecated recently or ever.

Any way, I have pushed a commit to officially deprecate the old name. In Numpy 1.25.0?

Reply:
> Since this is python, everything is public API.

The long-standing convention in Python is that a leading underscore means *private*.  Any code can access that function (so it is public in that sense), but developers should not consider it part of a stable API.

Reply:
So, no `DeprecationWarning` in your opinion? How to reach a consensus?

Reply:
Yes, that's my opinion.  But if no one else chimes in, do what @mattip says. 😄

Reply:
I relent. This is fine, let's put it in as-is. Sorry for the noise.

Reply:
OK @mattip, removing the DeprecationWarning I had just added then. Thank you both for looking into this PR.